File: balanced_dihedral.py
"""
Counting balanced subsets of the dihedral group D_k  (|D_k| = n = 2k).

A subset A of a group G is *balanced* when |AA| = |AA^{-1}|.

KEEP THIS IN A PLAIN .py FILE and import it from Sage:
Sage's preparser turns integer literals into Sage Integers, whose bitwise
operations are one to two orders of magnitude slower than machine ints.
Files imported with `from balanced_dihedral import ...` are NOT preparsed.

--------------------------------------------------------------------------
The reduction this code is built on
--------------------------------------------------------------------------
Write D_k = <rho, tau | rho^k = tau^2 = 1, tau rho tau = rho^{-1}> and split
a subset A into its rotations and its reflections:

    A = { rho^a : a in R }  u  { tau rho^t : t in T },     R, T subseteq Z_k.

From  rho^a rho^b = rho^{a+b},  rho^a (tau rho^b) = tau rho^{b-a},
      (tau rho^a) rho^b = tau rho^{a+b},  (tau rho^a)(tau rho^b) = rho^{b-a},
and (tau rho^a)^{-1} = tau rho^a, one gets, with all sums/differences taken
in Z_k,

    P = R+R,  D = R-R,  E = T-T,  M = T-R,  N = T+R

    AA      = { rho^x : x in P u E } u { tau rho^y : y in M u N }
    AA^{-1} = { rho^x : x in D u E } u { tau rho^y : y in M }

    A is balanced  <=>  |P u E| + |M u N| = |D u E| + |M|.

So the whole computation lives in Z_k and every set is a k-bit mask; sumsets
are unions of cyclic shifts, i.e. single machine words.

--------------------------------------------------------------------------
The three things that make this fast
--------------------------------------------------------------------------
1. Bitmasks.  |R| * |S| pairwise products become |R| shift-or's.
2. Incremental DFS.  Adding one element to R or T updates P, D, E, M, N with
   O(1) word operations, e.g. (R u {x}) + (R u {x}) = (R+R) u (x + (R u {x})).
3. Two subtree shortcuts, which is where the real asymptotic win is:
   (a) if P u E = D u E = M = Z_k then AA = AA^{-1} = G, and the same holds
       for *every* superset of A, so 2^rem subsets are counted at once;
   (b) |A| >= n/2 makes every strict superset balanced (backstop; rarely fires
       before (a) does).
   Cost drops from Theta(n^2 4^k) to about 3.15^k.
4. Aut(D_k) contains tau rho^t -> tau rho^{t+v}, so only pairs with 0 in T
   need enumerating:  #balanced with T nonempty = k * sum_{0 in T} 1/|T|.
   Another factor of ~2.7 (unsaturated sets have |T| ~ 0.35k on average,
   not k/2).
"""

from fractions import Fraction

__all__ = ["count_balanced_dihedral", "count_balanced_dihedral_naive",
           "dihedral_element_order_is_rotations_first"]


def count_balanced_dihedral(k):
    """Number of subsets A of D_k (order 2k) with |AA| = |AA^{-1}|."""
    k = int(k)
    full = (1 << k) - 1
    nslot = 2 * k - 1

    # slot table: (kind, v, w, bit, remT, 2^remR); reflection 0 is forced in.
    slots = []
    for i in range(k):
        slots.append((0, i, (k - i) % k, 1 << i))
        if i:
            slots.append((1, i, (k - i) % k, 1 << i))
    tab = []
    for j in range(nslot):
        rem = slots[j + 1:]
        rt = sum(1 for s in rem if s[0])
        tab.append(slots[j] + (rt, 1 << (len(rem) - rt)))

    cnt = [0] * (2 * k + 2)          # cnt[t] : balanced pairs with 0 in T, |T| = t
    acc = [[0] * (k + 2) for _ in range(k + 2)]   # pooled saturated subtrees

    def rec(idx, R, Rn, T, Tn, P, D, E, M, N, size, t):
        for j in range(idx, nslot):
            kind, v, w, bit, remT, p2 = tab[j]
            if kind == 0:
                R2 = R | bit
                Rn2 = Rn | (1 << w)
                P2 = P | (((R2 << v) | (R2 >> (k - v))) & full)
                D2 = D | (((Rn2 << v) | (Rn2 >> (k - v))) & full) \
                       | (((R2 << w) | (R2 >> (k - w))) & full)
                E2 = E
                M2 = M | (((T << w) | (T >> (k - w))) & full)
                N2 = N | (((T << v) | (T >> (k - v))) & full)
                T2, Tn2, t2 = T, Tn, t
            else:
                T2 = T | bit
                Tn2 = Tn | (1 << w)
                E2 = E | (((Tn2 << v) | (Tn2 >> (k - v))) & full) \
                       | (((T2 << w) | (T2 >> (k - w))) & full)
                M2 = M | (((Rn << v) | (Rn >> (k - v))) & full)
                N2 = N | (((R << v) | (R >> (k - v))) & full)
                R2, Rn2, P2, D2, t2 = R, Rn, P, D, t + 1
            PE = P2 | E2
            DE = D2 | E2
            if PE == full and DE == full and M2 == full:
                acc[remT][t2] += p2                 # AA = AA^{-1} = G, hereditary
                continue
            bal = (PE.bit_count() + (M2 | N2).bit_count()
                   == DE.bit_count() + M2.bit_count())
            if size + 1 >= k:                        # |A| >= n/2
                acc[remT][t2] += p2
                if not bal:
                    cnt[t2] -= 1
                continue
            if bal:
                cnt[t2] += 1
            if j + 1 < nslot:
                rec(j + 1, R2, Rn2, T2, Tn2, P2, D2, E2, M2, N2, size + 1, t2)

    cnt[1] += 1                                      # A = {tau}
    rec(0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1)

    # expand the pooled subtrees, then undo the 1/|T| weighting
    from math import comb
    tot = list(cnt)
    for rt in range(k + 1):
        for t in range(1, k + 1):
            a = acc[rt][t]
            if a:
                for e in range(rt + 1):
                    tot[t + e] += a * comb(rt, e)
    total = sum(Fraction(tot[t], t) for t in range(1, k + 1)) * k

    # T empty: A sits inside the rotation subgroup, |R+R| vs |R-R| in Z_k
    def rot_only(idx, R, Rn, P, D):
        c = 0
        for i in range(idx, k):
            w = (k - i) % k
            R2 = R | (1 << i)
            Rn2 = Rn | (1 << w)
            P2 = P | (((R2 << i) | (R2 >> (k - i))) & full)
            D2 = D | (((Rn2 << i) | (Rn2 >> (k - i))) & full) \
                   | (((R2 << w) | (R2 >> (k - w))) & full)
            if P2 == full and D2 == full:
                c += 1 << (k - i - 1)
                continue
            if P2.bit_count() == D2.bit_count():
                c += 1
            if i + 1 < k:
                c += rot_only(i + 1, R2, Rn2, P2, D2)
        return c

    total += 1 + rot_only(0, 0, 0, 0, 0)             # the empty set, and T = {}
    assert total.denominator == 1
    return int(total)


# ---------------------------------------------------------------- checking --
def count_balanced_dihedral_naive(k):
    """O(4^k k^2) reference: every subset, multiplication tables, no shortcuts."""
    k = int(k)
    n = 2 * k
    mul = [[0] * n for _ in range(n)]
    for a in range(n):
        for b in range(n):
            if a < k and b < k:    mul[a][b] = (a + b) % k
            elif a < k:            mul[a][b] = k + (b - k - a) % k
            elif b < k:            mul[a][b] = k + (a - k + b) % k
            else:                  mul[a][b] = (b - a) % k
    inv = [(-i) % k if i < k else i for i in range(n)]
    div = [[mul[a][inv[b]] for b in range(n)] for a in range(n)]
    total = 0
    for m in range(1 << n):
        A = [i for i in range(n) if m >> i & 1]
        if len({mul[x][y] for x in A for y in A}) == len({div[x][y] for x in A for y in A}):
            total += 1
    return total


def dihedral_element_order_is_rotations_first(G, k):
    """Run this in Sage before trusting any code that assumes list(G)[:k] are
    the rotations -- the original algorithm's AA^{-1} formula is only valid
    under that assumption, and Sage does not promise it."""
    els = list(G)
    return all(els[i].order() != 2 or els[i].is_one() for i in range(k)) and \
           len({els[i] * els[j] for i in range(k) for j in range(k)}) <= k


if __name__ == "__main__":
    import sys, time
    known = {1: 4, 2: 16, 3: 46, 4: 184, 5: 684, 6: 2830, 7: 11008, 8: 46584,
             9: 192250, 10: 806396, 11: 3328472, 12: 13788886, 13: 56881972,
             14: 233244664, 15: 957162366}
    sys.setrecursionlimit(10000)
    for k in range(1, 15):
        t = time.time(); v = count_balanced_dihedral(k); dt = time.time() - t
        tag = "OK" if known.get(k) == v else "??"
        print(f"k={k:2d}  {v:14d}  {dt:8.3f}s  {tag}", flush=True)

File: balanced_dihedral.c
/* ===========================================================================
 * Balanced subsets of the dihedral group D_k  (|D_k| = n = 2k),  k <= 31.
 * A is balanced iff |AA| = |AA^{-1}|.
 *
 *   A = {rho^a : a in R} u {tau rho^t : t in T},   R, T subseteq Z_k
 *   P = R+R,  D = R-R,  E = T-T,  M = T-R,  N = T+R      (all inside Z_k)
 *
 *   AA      = (P u E) rotations + (M u N) reflections
 *   AA^{-1} = (D u E) rotations +    M    reflections
 *   balanced  <=>  |P u E| + |M u N| = |D u E| + |M|
 *
 * Every set is a k-bit mask; a sumset is a union of cyclic shifts, and adding
 * one element to R or T updates all five masks in O(1) word operations.
 *
 * Two hereditary shortcuts collapse whole subtrees:
 *   - P u E = D u E = M = Z_k  =>  AA = AA^{-1} = G for A and all supersets
 *   - |A| >= n/2               =>  every strict superset is balanced
 * and Aut(D_k) > (tau rho^t -> tau rho^{t+v}) lets us enumerate only pairs
 * with 0 in T, undoing the weighting with
 *      #balanced (T nonempty) = k * sum_{0 in T, balanced} 1/|T|.
 * T = empty is handled by a separate 2^k pass (|R+R| vs |R-R| in Z_k).
 *
 * Build:  gcc -O3 -march=native -funroll-loops -o balanced_dihedral balanced_dihedral.c
 * Serial: ./balanced_dihedral K
 * Shard:  ./balanced_dihedral K NTASKS TASKID SPLITDEPTH     (see run_parallel.py)
 *         SPLITDEPTH ~ 6 gives millions of independent subtrees, round-robin
 *         over the tasks; every task repeats the (tiny) tree above that depth
 *         but only task 0 counts it.
 * ======================================================================== */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

typedef unsigned int u32;
typedef unsigned long long u64;
typedef unsigned __int128 u128;

static int K, NSLOT, SPLIT = 12, NTASK = 1, TASKID = 0;
static u32 FULL;
static int kind_[64], sv_[64], sw_[64], remR_[64], remT_[64];
static u32 bit_[64];
static u64 pow2remR_[64];
static __int128 cnt[64];     /* cnt[t] = #balanced pairs, 0 in T, |T| = t   */
static u128 acc[64][64];     /* acc[remT][t] = pooled 2^remR from pruned subtrees */
static u64 prefix_ctr;

static inline u32 rl(u32 m, int s) { return ((m << s) | (m >> (K - s))) & FULL; }
#define PC __builtin_popcount

/* ---------- part (ii): T nonempty, normalised so that 0 in T ------------- */
static void rec(int idx, u32 R, u32 Rn, u32 T, u32 Tn,
                u32 P, u32 D, u32 E, u32 M, u32 Nx, int size, int t)
{
    for (int j = idx; j < NSLOT; j++) {
        u32 R2, Rn2, T2, Tn2, P2, D2, E2, M2, N2;
        const int v = sv_[j], w = sw_[j];
        int t2 = t;
        if (kind_[j] == 0) {
            R2 = R | bit_[j];  Rn2 = Rn | (1u << w);  T2 = T;  Tn2 = Tn;
            P2 = P | rl(R2, v);
            D2 = D | rl(Rn2, v) | rl(R2, w);
            E2 = E;
            M2 = M | rl(T, w);
            N2 = Nx | rl(T, v);
        } else {
            T2 = T | bit_[j];  Tn2 = Tn | (1u << w);  R2 = R;  Rn2 = Rn;
            P2 = P;  D2 = D;
            E2 = E | rl(Tn2, v) | rl(T2, w);
            M2 = M | rl(Rn, v);
            N2 = Nx | rl(R, v);
            t2 = t + 1;
        }
        const u32 PE = P2 | E2, DE = D2 | E2;
        const int own = (size + 1 > SPLIT) | (TASKID == 0); /* above the split depth: task 0 only */
        if (PE == FULL && DE == FULL && M2 == FULL) {   /* AA = AA^-1 = G, hereditary */
            if (own) acc[remT_[j]][t2] += pow2remR_[j];
            continue;
        }
        const int bal = (PC(PE) + PC(M2 | N2) == PC(DE) + PC(M2));
        if (size + 1 >= K) {                            /* |A| >= n/2: supersets balanced */
            if (own) { acc[remT_[j]][t2] += pow2remR_[j]; cnt[t2] += bal ? 0 : -1; }
            continue;
        }
        if (own) cnt[t2] += bal;
        if (j + 1 < NSLOT) {
            if (size + 1 == SPLIT) { if (prefix_ctr++ % NTASK != (u64)TASKID) continue; }
            rec(j + 1, R2, Rn2, T2, Tn2, P2, D2, E2, M2, N2, size + 1, t2);
        }
    }
}

/* ---------- part (i): T empty, i.e. A inside the rotation subgroup ------- */
static u64 rot_only(int idx, u32 R, u32 Rn, u32 P, u32 D)
{
    u64 c = 0;
    for (int i = idx; i < K; i++) {
        u32 R2 = R | (1u << i), Rn2 = Rn | (1u << ((K - i) % K));
        u32 P2 = P | rl(R2, i), D2 = D | rl(Rn2, i) | rl(R2, (K - i) % K);
        if (P2 == FULL && D2 == FULL) { c += 1ULL << (K - i - 1); continue; }
        c += (PC(P2) == PC(D2));
        if (i + 1 < K) c += rot_only(i + 1, R2, Rn2, P2, D2);
    }
    return c;
}

int main(int argc, char **argv)
{
    K = argc > 1 ? atoi(argv[1]) : 8;
    NTASK  = argc > 2 ? atoi(argv[2]) : 1;
    TASKID = argc > 3 ? atoi(argv[3]) : 0;
    SPLIT  = argc > 4 ? atoi(argv[4]) : 12;
    if (NTASK == 1) SPLIT = 1 << 30;
    FULL = (1u << K) - 1;

    /* slot order: interleave rotations and reflections; reflection 0 is forced in */
    int p = 0;
    for (int i = 0; i < K; i++) {
        kind_[p] = 0; sv_[p] = i; sw_[p] = (K - i) % K; bit_[p] = 1u << i; p++;
        if (i) { kind_[p] = 1; sv_[p] = i; sw_[p] = (K - i) % K; bit_[p] = 1u << i; p++; }
    }
    NSLOT = p;                                     /* = 2k-1 */
    for (int j = 0; j < NSLOT; j++) {
        int rr = 0, rt = 0;
        for (int q = j + 1; q < NSLOT; q++) { if (kind_[q]) rt++; else rr++; }
        remR_[j] = rr; remT_[j] = rt; pow2remR_[j] = 1ULL << rr;
    }

    struct timespec a, b; clock_gettime(CLOCK_MONOTONIC, &a);
    memset(cnt, 0, sizeof cnt); memset(acc, 0, sizeof acc);
    /* start with T = {0}: E = {0}, everything else empty, |A| = 1 */
    if (TASKID == 0) cnt[1] += 1;                  /* A = {tau} itself is balanced */
    rec(0, 0, 0, 1u, 1u, 0, 0, 1u, 0, 0, 1, 1);
    u64 empty_T = (NTASK == 1 || TASKID == 0) ? 1 + rot_only(0, 0, 0, 0, 0) : 0;
    clock_gettime(CLOCK_MONOTONIC, &b);

    /* expand pooled subtree counts, C(remT, extra reflections) */
    static u128 C[64][64];
    for (int i = 0; i < 64; i++) { C[i][0] = 1; for (int j = 1; j <= i; j++) C[i][j] = C[i-1][j-1] + C[i-1][j]; }
    static __int128 tot[64];
    for (int t = 1; t <= K; t++) tot[t] = cnt[t];
    for (int rt = 0; rt <= K; rt++)
        for (int t = 1; t <= K; t++)
            if (acc[rt][t])
                for (int e = 0; e <= rt; e++) tot[t + e] += (__int128)(acc[rt][t] * C[rt][e]);

    /* answer = #(T empty) + k * sum_t tot[t]/t   (exact via a common denominator) */
    u128 L = 1;
    for (int t = 1; t <= K; t++) { u128 g = L, h = t; while (h) { u128 r = g % h; g = h; h = r; } L = L / g * t; }
    __int128 s = 0;
    for (int t = 1; t <= K; t++) s += tot[t] * (__int128)(L / t);
    __int128 ans = (__int128)K * s / (__int128)L + empty_T;

    double sec = (b.tv_sec-a.tv_sec) + 1e-9*(b.tv_nsec-a.tv_nsec);
    if (NTASK == 1) {
        printf("balanced subsets of D_%d = %llu\n", K, (u64)ans);
        printf("  k = %d   %.1f core-seconds\n", K, sec);
    } else {
        printf("k=%2d task %d/%d  partial: ", K, TASKID, NTASK);
        for (int t = 1; t <= K; t++) printf("%lld ", (long long)tot[t]);
        printf(" emptyT=%llu seconds=%.3f\n", empty_T, sec);
    }
    return 0;
}
